### ISSUE TITLE
[PT2] Return int32 indices in max_pool2d_with_indices

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#103785]
### 🚀 The feature, motivation and pitch

Another case of https://github.com/pytorch/pytorch/issues/103475. 

If we guard on the tensor having numel < 2^32, we can return int32 indices. On one max_pool2d_with_indices_backward kernel I tested in the backward, using int32 instead of int64 sped it up 27%. 

Related: https://github.com/pytorch/pytorch/issues/103111

Because we sometimes fallback to eager in lowering max_pool2d_backward, we'd either have to figure out a way not to fallback eager or add eager support for int32 in max_pool2d_backward.

cc @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @wconstab

**Comment by eellison** [comment]
Additionally each index has a known range of values. If the kernel size is 2x2, then for any particular element there is only 4 different values it could come from. The index could be represented with int2. For 3x3 kernel int4.

**Comment by eellison** [comment]
Additionally each index has a known range of values. If the kernel size is 2x2, then for any particular element there is only 4 different values it could come from. We could use int2 here to store the index. And int4 if the kernel size is 3x3 (double check my understanding of how max_pool2d_indices work - but I think this is right).

**Comment by abdullah599** [comment]
HI! I am new to Open-Source Contribution. I understand this optimizing problem and also your point about why we should use int2 for storing indices if the kernel size is 2x2 and so on. 
I am new to this. I'll be very grateful to anyone who could guide me on how and where can I perform the changes to resolve this issue.

**Comment by amjames** [comment]
The range for the index comes from the input size. The index is the location in the original HxW plane where the max value in the output came from. So an optimization like this is possible, but the dtype for the indices would be determined by the input shape, not the kernel size. 

See this example:

```python
>>> x = torch.arange(32).reshape(1, 2, 4, 4)
>>> torch.nn.functional.max_pool2d_with_indices(x, (2, 2))[-1]
 tensor([[[[ 5,  7],
           [13, 15]],
 
          [[ 5,  7],
           [13, 15]]]]))

```

The index values are in [0, 15], from the input shape (4, 4) not the kernel size of (2, 2).

**Comment by peterbell10** [comment]
@amjames yes but the indices for a particular element in the output can only correspond to one of the elements used in the pooling kernel at that point, so you could efficiently represent it as a linear index into that kernel. This would give a different result from the eager kernel, but as long as it's only used for the backward formula this is fine and means we save less data for the backward pass.

**Comment by eellison** [comment]
@amjames I partially implemented this but didn't fully land: 

https://github.com/pytorch/pytorch/pull/105687/

Just needs a faster max_pool2d_backward triton implemenation

**Comment by amjames** [comment]
Ah, I missed that detail about the indices having a different interpretation in this context.

**Comment by amjames** [comment]
@eellison are you looking for someone to pick up that PR?

**Comment by amjames** [comment]
cleaning up my backlog @eellison can we consider this resolved via #122832 ? IIRC actually changing the return type has some wider implications that will be difficult, but we now allow an optimization in compile where smaller offsets are stored for backward and the full index is recomputed.

**Comment by eellison** [comment]
Yea, let's close it
